### PR TITLE
Don't force a node-reassigment when recording a new value of keys_changed_at

### PR DESCRIPTION
**Pull request by rfk**
During the initial rollout of keys_changed_at, we may be in a situation where the tokenserver first sees a new value for keys_changed_at after some older version of the server has already processed the node-reassignment for the corresponding change in client_state. We don't want to force users through
another node-reassignment in that case, since they would already have done one.

This changes the sql backend to only force a node-reassignment when client_state changes, and adds some tests to ensure it. The memory backend was accidentally doing the right thing already, although the tests uncovered an issue with handing out mutable references to its internal state that I've had to fix as part of this PR.

### CODE REVIEW

**Comment by jrgm** [comment]
Say @rfk - if you push up a `dockerpush.release-1.5.1-avoid-node-reassignment` branch I can deploy that to stage now for you to confirm this change on stage.

**Comment by rfk** [comment]
👍 pushing now under that name

**Comment by jrgm** [comment]
Stage is now commit  f5c32e69f
```
$  curl -s https://token.stage.mozaws.net/__version__ | python -mjson.tool
{
    "build": "https://circleci.com/gh/mozilla-services/tokenserver/295",
    "commit": "f5c32e69fb69f7bbc23167ce63263b486ff11d9d",
    "source": "https://github.com/mozilla-services/tokenserver",
    "version": ""
}
```

**Comment by rfk** [comment]
Thanks! To help with testing, whenever you have a chance, could you please `UPDATE users SET keys_changed_at = NULL WHERE uid = 5469307` on stage? That's my current user account, and will help me test upgrading from an existing record without the keys-changed-at value set.

**Comment by jrgm** [comment]
> could you please UPDATE users SET keys_changed_at = NULL WHERE uid = 5469307 on stage? 

Done.
```
> select keys_changed_at from users where uid = 5469307;
+-----------------+
| keys_changed_at |
+-----------------+
|            NULL |
+-----------------+
```

**Comment by rfk** [comment]
Thanks! I just ran my test again, did it set `keys_changed_at` for this `uid`? (as opposed to the previous behaviour, where it would have created a new `uid` and marked the existing one as replaced).

**Comment by jrgm** [comment]
```
> select keys_changed_at from users where uid = 5469307;
+-----------------+
| keys_changed_at |
+-----------------+
|   1575525631477 |
+-----------------+
```

**Comment by rfk** [comment]
Awesome, looks like this is working as intended 👍

**Comment by jrgm** [comment]
So, when there's a tag for this should I try this in production tomorrow morning PST, or wait for your morning? (Note: I have an appointment tomorrow that I cannot reschedule again at 2300Z)

**Comment by jrgm** [comment]
but I'd really like to get this done.

**Comment by rfk** [comment]
I don't want to send this prod without an r? from @pjenvey, but perhaps you can coordinate with him for a tag tomorrow morning. I'm happy that it's behaving as I expect.

File: tokenserver/assignment/sqlnode/sql.py [hunk 1]
```diff
@@ -65,14 +65,18 @@
...
 update
     users
 set
-    generation = :generation
+    generation = COALESCE(:generation, generation),
+    keys_changed_at = COALESCE(:keys_changed_at, keys_changed_at)
 where
     service = :service and email = :email and
-    generation < :generation and replaced_at is null
+    generation <= COALESCE(:generation, generation) and
```
Comment (pjenvey) [inline tokenserver/assignment/sqlnode/sql.py]: I had to stare at these bits for a while. Keeping this logic in Python (probably around an actual sqlalchemy update expression) would help readability, but I'm fine committing it as is for now.

Comment (rfk) [reply tokenserver/assignment/sqlnode/sql.py]: This is mostly belt-and-braces stuff to ensure that we don't move any timestamps backwards, I'll add more of a comment in the statement in the meantime.

**Pull request merged**